File: backend/src/main/java/com/unbora/api/domain/user/UsersService.java
package com.unbora.api.domain.user;

import com.unbora.api.common.PasswordUtil;
import com.unbora.api.common.exception.ApiException;
import com.unbora.api.domain.user.dto.*;
import com.unbora.api.kafka.KafkaEventPublisher;
import com.unbora.api.kafka.event.UserActivityEvent;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Service
public class UsersService {

    private final UserRepository userRepository;
    private final KafkaEventPublisher kafkaEventPublisher;
    private final GoogleAuthService googleAuthService;

    public UsersService(
            UserRepository userRepository,
            KafkaEventPublisher kafkaEventPublisher,
            GoogleAuthService googleAuthService
    ) {
        this.userRepository = userRepository;
        this.kafkaEventPublisher = kafkaEventPublisher;
        this.googleAuthService = googleAuthService;
    }

    public PublicUserDto toPublic(User user) {
        return new PublicUserDto(
                user.getId(),
                user.getName(),
                user.getEmail(),
                user.getGuest(),
                user.getPlatform(),
                user.getRole(),
                user.getBusinessName(),
                user.getCreatedAt().toString(),
                user.getLastSeenAt().toString()
        );
    }

    public List<PublicUserDto> list() {
        return userRepository.findAllByOrderByLastSeenAtDesc().stream()
                .map(this::toPublic)
                .toList();
    }

    public UserStatsDto stats() {
        long total = userRepository.count();
        long guests = userRepository.countByIsGuestTrue();
        long registered = userRepository.countByIsGuestFalse();

        Instant startOfToday = LocalDate.now(ZoneId.of("America/Fortaleza"))
                .atStartOfDay(ZoneId.of("America/Fortaleza"))
                .toInstant();

        long activeToday = userRepository.findAll().stream()
                .filter(u -> u.getLastSeenAt() != null && u.getLastSeenAt().isAfter(startOfToday))
                .count();

        return new UserStatsDto(total, guests, registered, activeToday);
    }

    @Transactional
    public PublicUserDto register(RegisterUserDto dto) {
        String email = dto.email().trim().toLowerCase();

        boolean exists = userRepository.findByEmail(email)
                .filter(u -> !Boolean.TRUE.equals(u.getGuest()))
                .isPresent();

        if (exists) {
            throw new ApiException(HttpStatus.CONFLICT, "Este e-mail já está cadastrado.");
        }

        User user = new User();
        user.setId(UUID.randomUUID().toString());
        user.setName(dto.name().trim());
        user.setEmail(email);
        user.setGuest(false);
        user.setPasswordHash(PasswordUtil.hashPassword(dto.password()));
        user.setPlatform(dto.platform());
        user.setRole("user");
        user.setCreatedAt(Instant.now());
        user.setLastSeenAt(Instant.now());

        PublicUserDto saved = toPublic(userRepository.save(user));

        kafkaEventPublisher.publishUserActivity(new UserActivityEvent(
                "USER_REGISTERED",
                saved.id(),
                saved.email(),
                saved.role(),
                saved.platform(),
                Instant.now(),
                Map.of("name", saved.name())
        ));

        return saved;
    }

    @Transactional
    public PublicUserDto registerMerchant(RegisterMerchantDto dto) {
        String email = dto.email().trim().toLowerCase();

        boolean exists = userRepository.findByEmail(email)
                .filter(u -> !Boolean.TRUE.equals(u.getGuest()))
                .isPresent();

        if (exists) {
            throw new ApiException(HttpStatus.CONFLICT, "Este e-mail já está cadastrado.");
        }

        User user = new User();
        user.setId(UUID.randomUUID().toString());
        user.setName(dto.name().trim());
        user.setEmail(email);
        user.setGuest(false);
        user.setPasswordHash(PasswordUtil.hashPassword(dto.password()));
        user.setPlatform(dto.platform());
        user.setRole("merchant");
        user.setBusinessName(dto.businessName().trim());
        user.setCreatedAt(Instant.now());
        user.setLastSeenAt(Instant.now());

        PublicUserDto saved = toPublic(userRepository.save(user));

        kafkaEventPublisher.publishUserActivity(new UserActivityEvent(
                "MERCHANT_REGISTERED",
                saved.id(),
                saved.email(),
                saved.role(),
                saved.platform(),
                Instant.now(),
                Map.of("businessName", dto.businessName())
        ));

        return saved;
    }

    @Transactional
    public PublicUserDto login(LoginUserDto dto) {
        String email = dto.email().trim().toLowerCase();

        User user = userRepository.findByEmail(email)
                .filter(u -> !Boolean.TRUE.equals(u.getGuest()))
                .orElseThrow(() -> new ApiException(HttpStatus.UNAUTHORIZED, "E-mail ou senha inválidos."));

        if (!PasswordUtil.verify(dto.password(), user.getPasswordHash())) {
            throw new ApiException(HttpStatus.UNAUTHORIZED, "E-mail ou senha inválidos.");
        }

        user.setLastSeenAt(Instant.now());
        PublicUserDto saved = toPublic(userRepository.save(user));

        kafkaEventPublisher.publishUserActivity(new UserActivityEvent(
                "USER_LOGGED_IN",
                saved.id(),
                saved.email(),
                saved.role(),
                saved.platform(),
                Instant.now(),
                Map.of()
        ));

        return saved;
    }

    @Transactional
    public PublicUserDto googleLogin(GoogleLoginDto dto) {
        String email;
        String name;
        String googleId;
        String platform = dto.platform();

        if (dto.idToken() != null && !dto.idToken().isBlank()) {
            GoogleAuthService.GoogleUserProfile profile = googleAuthService.verifyIdToken(dto.idToken());
            email = profile.email();
            name = profile.name();
            googleId = profile.googleId();
        } else if (dto.email() != null && !dto.email().isBlank()) {
            email = dto.email().trim().toLowerCase();
            name = (dto.name() != null && !dto.name().isBlank()) ? dto.name().trim() : "Usuário Google";
            googleId = (dto.googleId() != null && !dto.googleId().isBlank()) ? dto.googleId() : UUID.randomUUID().toString();
        } else {
            throw ApiException.badRequest("Credenciais de login Google ausentes.");
        }

        User user = userRepository.findByEmail(email)
                .filter(u -> !Boolean.TRUE.equals(u.getGuest()))
                .orElse(null);

        if (user != null) {
            user.setLastSeenAt(Instant.now());
            if (name != null && !name.isBlank() && ("Usuário".equalsIgnoreCase(user.getName()) || user.getName().isBlank())) {
                user.setName(name);
            }
            if (platform != null) user.setPlatform(platform);
            PublicUserDto saved = toPublic(userRepository.save(user));

            kafkaEventPublisher.publishUserActivity(new UserActivityEvent(
                    "GOOGLE_LOGIN",
                    saved.id(),
                    saved.email(),
                    saved.role(),
                    saved.platform(),
                    Instant.now(),
                    Map.of("googleId", googleId)
            ));

            return saved;
        }

        User newUser = new User();
        newUser.setId(UUID.randomUUID().toString());
        newUser.setName(name);
        newUser.setEmail(email);
        newUser.setGuest(false);
        newUser.setPlatform(platform);
        newUser.setRole("user");
        newUser.setCreatedAt(Instant.now());
        newUser.setLastSeenAt(Instant.now());

        PublicUserDto saved = toPublic(userRepository.save(newUser));

        kafkaEventPublisher.publishUserActivity(new UserActivityEvent(
                "GOOGLE_REGISTER",
                saved.id(),
                saved.email(),
                saved.role(),
                saved.platform(),
                Instant.now(),
                Map.of("googleId", googleId)
        ));

        return saved;
    }

    @Transactional
    public PublicUserDto sync(SyncUserDto dto) {
        String email = dto.email() != null ? dto.email().trim().toLowerCase() : "";

        User user = userRepository.findById(dto.id()).orElse(null);

        if (user == null && !email.isBlank()) {
            user = userRepository.findByEmail(email)
                    .filter(u -> !Boolean.TRUE.equals(u.getGuest()))
                    .orElse(null);
        }

        if (user == null) {
            User newUser = new User();
            newUser.setId(dto.id());
            newUser.setName(dto.name().trim());
            newUser.setEmail(email);
            newUser.setGuest(dto.isGuest());
            newUser.setPlatform(dto.platform());
            newUser.setRole("user");
            newUser.setCreatedAt(Instant.now());
            newUser.setLastSeenAt(Instant.now());

            PublicUserDto saved = toPublic(userRepository.save(newUser));

            kafkaEventPublisher.publishUserActivity(new UserActivityEvent(
                    "USER_SYNC_NEW",
                    saved.id(),
                    saved.email(),
                    saved.role(),
                    saved.platform(),
                    Instant.now(),
                    Map.of("isGuest", saved.isGuest())
            ));

            return saved;
        }

        user.setName(dto.name().trim());
        if (!email.isBlank()) {
            user.setEmail(email);
        }
        user.setGuest(dto.isGuest());
        if (dto.platform() != null) {
            user.setPlatform(dto.platform());
        }
        user.setLastSeenAt(Instant.now());

        PublicUserDto saved = toPublic(userRepository.save(user));

        kafkaEventPublisher.publishUserActivity(new UserActivityEvent(
                "USER_SYNCED",
                saved.id(),
                saved.email(),
                saved.role(),
                saved.platform(),
                Instant.now(),
                Map.of("isGuest", saved.isGuest())
        ));

        return saved;
    }

    public PublicUserDto findById(String id) {
        return userRepository.findById(id)
                .map(this::toPublic)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Usuário não encontrado."));
    }
}

File: backend-nest-backup/src/users/users.service.ts
import {
  ConflictException,
  Injectable,
  NotFoundException,
  UnauthorizedException,
} from '@nestjs/common';
import { createHash, randomUUID } from 'crypto';

import { PrismaService } from '../prisma/prisma.service';

import {
  GoogleLoginDto,
  LoginUserDto,
  PublicUser,
  RegisterMerchantDto,
  RegisterUserDto,
  SyncUserDto,
  UserStats,
} from './users.types';

@Injectable()
export class UsersService {
  constructor(private readonly prisma: PrismaService) {}

  private hashPassword(password: string): string {
    return createHash('sha256').update(password).digest('hex');
  }

  private toPublic(user: {
    id: string;
    name: string;
    email: string;
    isGuest: boolean;
    platform: string | null;
    role: 'user' | 'merchant';
    businessName: string | null;
    createdAt: Date;
    lastSeenAt: Date;
  }): PublicUser {
    return {
      id: user.id,
      name: user.name,
      email: user.email,
      isGuest: user.isGuest,
      platform: user.platform ?? undefined,
      role: user.role,
      businessName: user.businessName ?? undefined,
      createdAt: user.createdAt.toISOString(),
      lastSeenAt: user.lastSeenAt.toISOString(),
    };
  }

  async list(): Promise<PublicUser[]> {
    const users = await this.prisma.user.findMany({
      orderBy: { lastSeenAt: 'desc' },
    });
    return users.map((user) => this.toPublic(user));
  }

  async stats(): Promise<UserStats> {
    const today = new Date();
    today.setHours(0, 0, 0, 0);

    const [total, guests, registered, activeToday] = await Promise.all([
      this.prisma.user.count(),
      this.prisma.user.count({ where: { isGuest: true } }),
      this.prisma.user.count({ where: { isGuest: false } }),
      this.prisma.user.count({ where: { lastSeenAt: { gte: today } } }),
    ]);

    return { total, guests, registered, activeToday };
  }

  async register(dto: RegisterUserDto): Promise<PublicUser> {
    const email = dto.email.trim().toLowerCase();

    const existing = await this.prisma.user.findFirst({
      where: { email, isGuest: false },
    });
    if (existing) {
      throw new ConflictException('Este e-mail já está cadastrado.');
    }

    const user = await this.prisma.user.create({
      data: {
        id: randomUUID(),
        name: dto.name.trim(),
        email,
        isGuest: false,
        passwordHash: this.hashPassword(dto.password),
        platform: dto.platform,
      },
    });

    return this.toPublic(user);
  }

  async registerMerchant(dto: RegisterMerchantDto): Promise<PublicUser> {
    const email = dto.email.trim().toLowerCase();

    const existing = await this.prisma.user.findFirst({
      where: { email, isGuest: false },
    });
    if (existing) {
      throw new ConflictException('Este e-mail já está cadastrado.');
    }

    const user = await this.prisma.user.create({
      data: {
        id: randomUUID(),
        name: dto.name.trim(),
        email,
        isGuest: false,
        passwordHash: this.hashPassword(dto.password),
        platform: dto.platform,
        role: 'merchant',
        businessName: dto.businessName.trim(),
      },
    });

    return this.toPublic(user);
  }

  async login(dto: LoginUserDto): Promise<PublicUser> {
    const email = dto.email.trim().toLowerCase();
    const user = await this.prisma.user.findFirst({
      where: { email, isGuest: false },
    });

    if (!user || user.passwordHash !== this.hashPassword(dto.password)) {
      throw new UnauthorizedException('E-mail ou senha inválidos.');
    }

    const updated = await this.prisma.user.update({
      where: { id: user.id },
      data: { lastSeenAt: new Date() },
    });

    return this.toPublic(updated);
  }

  async googleLogin(dto: GoogleLoginDto): Promise<PublicUser> {
    const email = dto.email.trim().toLowerCase();
    const existing = await this.prisma.user.findFirst({
      where: { email, isGuest: false },
    });

    const user = existing
      ? await this.prisma.user.update({
          where: { id: existing.id },
          data: {
            lastSeenAt: new Date(),
            platform: dto.platform ?? existing.platform,
          },
        })
      : await this.prisma.user.create({
          data: {
            id: randomUUID(),
            name: dto.name.trim(),
            email,
            isGuest: false,
            platform: dto.platform,
          },
        });

    return this.toPublic(user);
  }

  async sync(dto: SyncUserDto): Promise<PublicUser> {
    const email = dto.email?.trim().toLowerCase() ?? '';
    let user = await this.prisma.user.findUnique({ where: { id: dto.id } });

    if (!user && email) {
      user = await this.prisma.user.findFirst({
        where: { email, isGuest: false },
      });
    }

    if (!user) {
      user = await this.prisma.user.create({
        data: {
          id: dto.id,
          name: dto.name.trim(),
          email,
          isGuest: dto.isGuest,
          platform: dto.platform,
        },
      });
      return this.toPublic(user);
    }

    const updated = await this.prisma.user.update({
      where: { id: user.id },
      data: {
        name: dto.name.trim(),
        email: email || user.email,
        isGuest: dto.isGuest,
        platform: dto.platform ?? user.platform,
        lastSeenAt: new Date(),
      },
    });

    return this.toPublic(updated);
  }

  async findById(id: string): Promise<PublicUser> {
    const user = await this.prisma.user.findUnique({ where: { id } });
    if (!user) throw new NotFoundException('Usuário não encontrado.');
    return this.toPublic(user);
  }
}
